File: apps/front/src/models/season.ts
export type Season = {
  id: string
  // The same in every language.
  name: string
  // Missing for a season that is not shown on the seasons page.
  previewImage?: string
  startDate: number
  endDate?: number
  publicAddress?: string
  isActive: boolean
  isPrimary: boolean
  // Whether the site knows who is online in the season: it is running and has a server.
  onlineAvailable: boolean
  preregistration: boolean
  freeRegistration: boolean
  // The client version needed to join, e.g. "1.21.1".
  gameVersion?: string
  // An absolute link to the world archive of a finished season.
  worldUrl?: string
}

export type AdminSeason = Season & {
  // host:port of the shell service that serves the season server.
  shellAddress?: string
  // Plan web interface of the season server.
  planUrl?: string
}

export type SaveSeason = {
  name: string
  previewImage?: string
  startDate: string
  endDate?: string
  publicAddress?: string
  shellAddress?: string
  planUrl?: string
  isActive: boolean
  isPrimary: boolean
  preregistration: boolean
  freeRegistration: boolean
  gameVersion?: string
  worldUrl?: string
}

// One server of the season network: survival, farms, creative. Each has its own map and chunk claims.
export type SeasonWorld = {
  id: string
  seasonId: string
  // The key of the world page, /worlds/<slug>; unique within the season.
  slug: string
  // The same in every language.
  name: string
  // s3://bucket/key, the same form as Season.previewImage.
  previewImage?: string
  // The squaremap of the world server. The world has no map page without it.
  mapUrl?: string
  // Chunks one profile may claim in the world, over its claim dimensions together.
  claimLimit: number
  // squaremap world names where chunks can be claimed; empty for a view-only map.
  claimDimensions: string[]
  position: number
}

export type SaveSeasonWorld = {
  slug: string
  name: string
  previewImage?: string
  mapUrl?: string
  claimLimit: number
  claimDimensions: string[]
  position: number
}

export type ScreenshotAuthor = {
  id: string
  username: string
}

export type SeasonScreenshot = {
  id: string
  // s3://bucket/key, the same form as Season.previewImage.
  image: string
  title?: string
  position: number
  authors: ScreenshotAuthor[]
}

export type SaveSeasonScreenshot = {
  image: string
  title?: string
  position: number
  authorProfileIds: string[]
}

File: apps/api/internal/transport/http/responses/season.go
package responses

import "github.com/google/uuid"

type Season struct {
	ID uuid.UUID `json:"id"`
	// Name is the same in every language.
	Name string `json:"name"`
	// PreviewImage is missing for a season that is not shown on the seasons page.
	PreviewImage  *string `json:"previewImage,omitempty"`
	StartDate     int64   `json:"startDate"`
	EndDate       *int64  `json:"endDate,omitempty"`
	PublicAddress *string `json:"publicAddress,omitempty"`
	IsActive      bool    `json:"isActive"`
	IsPrimary     bool    `json:"isPrimary"`
	// OnlineAvailable tells whether the API knows who is online in the season: it is active and has a server.
	OnlineAvailable  bool    `json:"onlineAvailable"`
	Preregistration  bool    `json:"preregistration"`
	FreeRegistration bool    `json:"freeRegistration"`
	GameVersion      *string `json:"gameVersion,omitempty"`
	WorldURL         *string `json:"worldUrl,omitempty"`
	// MapURL is the squaremap of the season server, missing when chunk claims are off.
	MapURL     *string `json:"mapUrl,omitempty"`
	ClaimLimit int     `json:"claimLimit"`
}

type AdminSeason struct {
	Season
	ShellAddress *string `json:"shellAddress,omitempty"`
	PlanURL      *string `json:"planUrl,omitempty"`
}
